tools/migration/migrations/repo-1-2/migration.go

```diff
@@ -1 +1,376 @@
+	bb, err := chainStore.Ds.Get(headKey)
+	if err != nil {
+		return emptyCidSet, errors.Wrap(err, "failed to read headKey")
+	}
+
+	var cids types.SortedCidSet
+	if asCBOR {
+		err = cbor.DecodeInto(bb, &cids)
+
+	} else {
+		err = json.Unmarshal(bb, &cids)
+	}
+	if err != nil {
+		return emptyCidSet, errors.Wrap(err, "failed to cast headCids")
+	}
+
+	return cids, nil
+}
+
+// function wrapper for loading state root as JSON
+func loadStateRootAsJSON(ts types.TipSet, chainStore *migrationChainStore) (cid.Cid, error) {
+	return loadStateRoot(ts, false, chainStore)
+}
+
+// loadStateRoot loads the chain store metadata into chainStore, updating its
+// state root and then returning the state root + any error
+// pass true to load as CBOR (new format) or false to load as JSON (old format)
+func loadStateRoot(ts types.TipSet, asCBOR bool, chainStore *migrationChainStore) (cid.Cid, error) {
+	h, err := ts.Height()
+	if err != nil {
+		return cid.Undef, err
+	}
+	key := datastore.NewKey(makeKey(ts.String(), h))
+	bb, err := chainStore.Ds.Get(key)
+	if err != nil {
+		return cid.Undef, errors.Wrapf(err, "failed to read tipset key %s", ts.String())
+	}
+
+	var stateRoot cid.Cid
+	if asCBOR {
+		err = cbor.DecodeInto(bb, &stateRoot)
+	} else {
+		err = json.Unmarshal(bb, &stateRoot)
+	}
+	if err != nil {
+		return cid.Undef, errors.Wrapf(err, "failed to cast state root of tipset %s", ts.String())
+	}
+	return stateRoot, nil
+}
+
+// compareChainStores loads each chain store and iterates through each, comparing heads
+// then state root at each iteration, returning error if:
+//     * at any point state roots are not equal
+//     * the old and new chain store iterators are not complete at the same time
+// 	   * at the end the two heads are not equal
+func compareChainStores(ctx context.Context, oldChainStore *migrationChainStore, newChainStore *migrationChainStore) error {
+	oldTipCids, err := loadChainHeadAsJSON(oldChainStore)
+	if err != nil {
+		return err
+	}
+
+	newTipCids, err := loadChainHeadAsCBOR(newChainStore)
+	if err != nil {
+		return err
+	}
+
+	oldHeadTs, err := loadBlocksToChainStore(ctx, oldTipCids, oldChainStore)
+	if err != nil {
+		return err
+	}
+
+	newHeadTs, err := loadBlocksToChainStore(ctx, newTipCids, newChainStore)
+	if err != nil {
+		return err
+	}
+
+	if !newHeadTs.Equals(oldHeadTs) {
+		return errors.New("new and old head tipsets not equal")
+	}
+
+	oldIt := chain.IterAncestors(ctx, newChainStore, newHeadTs)
+	for newIt := chain.IterAncestors(ctx, oldChainStore, oldHeadTs); !newIt.Complete(); err = newIt.Next() {
+		if err != nil {
+			return err
+		}
+
+		newStateRoot, err := loadStateRoot(newIt.Value(), true, newChainStore)
+		if err != nil {
+			return err
+		}
+
+		oldStateRoot, err := loadStateRootAsJSON(oldIt.Value(), oldChainStore)
+		if err != nil {
+			return err
+		}
+		if !newStateRoot.Equals(oldStateRoot) {
+			return errors.New("current state root not equal for block")
+		}
+		if err = oldIt.Next(); err != nil {
```

Comment:
Because of the way [this iterator works](https://github.com/filecoin-project/go-filecoin/blob/master/chain/traversal.go#L17) if the old chain is somehow shorter than the new one then this loop will end up operating on an undefined tipset and raise a more cryptic error at line 329.  It would be a bit nicer to check `oldIt.Complete` at the top of the loop and explicitly return "old chain store is shorter than new chain store" there.

Reply:
ok, good idea